### ISSUE TITLE
CS0023 on Null-coalescing operator for generic type.

### ISSUE THREAD

**Issue by jeremyVignelles** [dotnet/roslyn#49652]
**Version Used**: 
dotnet SDK 5.0.100 on windows

**Steps to Reproduce**:
Clone this branch of the project : https://github.com/jeremyVignelles/repro-dotnet/tree/repro/null-coalescing-bug

Relevant part:

```cs
    public class TestClass<T> {
        public void DoStuff() {
            string p = "";
            var result = ((TestClass<T>?)null)?.DoOtherStuff(ref p);
        }


        public T? DoOtherStuff(ref string p) {
            return default;
        }
    }
```

**Expected Behavior**:
No build error. Especially, it shouldn't warn about type T when the left-hand side is of type `TestClass<T>`

**Actual Behavior**:
> D:\Projects\Tests\repro-dotnet\Program.cs(8,47): error CS0023: Operator '?' cannot be applied to operand of type 'T' [D:\Projects\Tests\repro-dotnet\repro-dotnet.csproj]

On line `var result = `

If I remove the `?`, like this:

```cs
 var result = ((TestClass<T>?)null).DoOtherStuff(ref p);
```

The compiler correctly indicates that the left part may be nullable.

Removing the `var result =` make the issue disappear.

**Workaround**
```cs
            var c = ((TestClass<T>?)null);
            if (c is not null) {
                var result = c.DoOtherStuff(ref p);
            }
```

**details**

<details>
<summary> dotnet --info </summary>

```
$ dotnet --info
SDK .NET (reflétant tous les fichiers global.json) :
 Version:   5.0.100
 Commit:    5044b93829

Environnement d'exécution :
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.100\

Host (useful for support):
  Version: 5.0.0
  Commit:  cf258a14b7

.NET SDKs installed:
  2.1.802 [C:\Program Files\dotnet\sdk]
  3.1.302 [C:\Program Files\dotnet\sdk]
  5.0.100 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download
```
</details>

**Comment by RikkiGibson** [comment]
Going by the [?. operator spec](https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#null-conditional-operator) this appears to be by-design. If `c.DoOtherStuff(ref p)` is not known to be of a reference type or value type, then there is no reasonable type we can choose for `c?.DoOtherStuff(ref p)`. Such an expression is only allowed when its value is not used.

The diagnostic here could perhaps be improved.

**Comment by CyrusNajmabadi** [comment]
We now report this as `'T' cannot be made nullable.` which is appropriate here.
